Sources/Managers/ApiManager.swift

```diff
@@ -18,43 +20,73 @@ class ApiManager {
         }
     }
 
-    func getUrlRequest(searchTerm: String) throws -> URLRequest? {
-        var urlComponents = URLComponents()
-        urlComponents.scheme = "https"
-        urlComponents.host = "api.github.com"
-        urlComponents.path = "/search/repositories"
-        urlComponents.queryItems = [
-            URLQueryItem(name: "q", value: searchTerm)
-        ]
-
-        guard let url = urlComponents.url else {
-            throw ApiError.invalidURL
-        }
-
-        return URLRequest(url: url)
-    }
-
     var decoder: JSONDecoder {
         let decoder = JSONDecoder()
         decoder.keyDecodingStrategy = .convertFromSnakeCase
         decoder.dateDecodingStrategy = .iso8601
         return decoder
     }
 
-    func getRepositories(for search: String) async throws -> [Repository] {
-        guard let urlRequest = try getUrlRequest(searchTerm: search)
+    func fetch<T: Codable>(
+        type: T.Type,
+        endpoint: Endpoint
+    ) async throws -> T {
+        let (data, response) = try await URLSession.shared
+            .data(for: endpoint.request)
+        guard (response as? HTTPURLResponse)?.statusCode == 200
         else {
             throw ApiError.badURLRequest
         }
-        let (data, response) = try await URLSession.shared.data(for: urlRequest)
+        do {
+            return try ApiManager.shared.decoder.decode(type, from: data)
```

Comment:
No need to fetch it through `ApiManager.shared` since you are already in this class. Just `decoder.decode()` will be enough.